bhttp/src/http_compat/decode.rs

```diff
@@ -0,0 +1,538 @@
+    pub fn into_full_message(self) -> Res<HttpMessage<R>> {
+        match self.parts {
+            HttpMessageParts::Request(parts) => Ok(HttpMessage::Request(Request::from_parts(
+                parts,
+                BhttpBody::new(self.async_message),
+            ))),
+            HttpMessageParts::Response(parts) => Ok(HttpMessage::Response(Response::from_parts(
+                parts,
+                BhttpBody::new(self.async_message),
+            ))),
+        }
+    }
+}
+
+impl TryFrom<crate::Header> for HttpMessageParts {
+    type Error = Error;
+
+    /// Convert a BHTTP header into either HTTP request or response parts
+    ///
+    /// This method parses the control data and field section from a BHTTP header
+    /// and constructs the corresponding HTTP request or response parts.
+    ///
+    /// # Arguments
+    ///
+    /// * `header` - The BHTTP header to convert
+    ///
+    /// # Returns
+    ///
+    /// Either HTTP request parts or response parts, depending on the control data,
+    /// or an error if parsing fails.
+    fn try_from(header: crate::Header) -> Res<Self> {
+        if header.control().is_request() {
+            Ok(HttpMessageParts::Request(request::Parts::try_from(header)?))
+        } else {
+            Ok(HttpMessageParts::Response(response::Parts::try_from(
+                header,
+            )?))
+        }
+    }
+}
+
+impl TryFrom<crate::Header> for request::Parts {
+    type Error = Error;
+
+    fn try_from(header: crate::Header) -> Res<Self> {
+        let control_data = header.control();
+
+        // Build the HTTP request parts
+        let mut builder = request::Builder::new();
+
+        // Set method
+        if let Some(method) = control_data.method() {
+            builder = builder.method(method);
+        }
+
+        // Convert field section to headers
+        let headers = HeaderMap::try_from(&header.fields)?;
+
+        // Set URI
+        let uri = {
+            let mut uri_builder = Uri::builder();
+
+            if let Some(scheme) = control_data.scheme() {
+                if !scheme.is_empty() {
+                    uri_builder = uri_builder.scheme(scheme);
+                }
+            }
+
+            // Handle authority field
+            match control_data.authority() {
+                Some(authority_bytes) if !authority_bytes.is_empty() => {
+                    uri_builder = uri_builder.authority(std::str::from_utf8(authority_bytes)?);
+                }
+                _ => {
+                    // The http::Uri does not allow an absent authority when the scheme is present, but
+                    // this is permitted in bhttp and report an InvalidUri(InvalidFormat) error. To
+                    // reconcile this difference, we will attempt to retrieve the authority from the
+                    // request headers.
+                    if let Some(scheme) = control_data.scheme() {
+                        if !scheme.is_empty() {
```

Comment:
Not sure why you are pulling the scheme again.  That should already be settled.

If your goal is to use `host` as a backup, I would do this:

```rust
let Some(auth) = control_data.authority() else {
    return Err(Error::Something);
};
// Add your comment here.
let authority = if auth.is_empty() {
    let Some(host) = headers.get("host") else {
        return Err(Error::Something);
    };
} else {
    auth;
};
uri_builder.authority(authority.to_str()?);
```